diff --git a/src/lib/optimizer/optimizer.cpp b/src/lib/optimizer/optimizer.cpp
index 916d2dbb6..193088db3 100644
--- a/src/lib/optimizer/optimizer.cpp
+++ b/src/lib/optimizer/optimizer.cpp
@@ -92,7 +92,7 @@ namespace opossum {
 std::shared_ptr<Optimizer> Optimizer::create_default_optimizer() {
   const auto optimizer = std::make_shared<Optimizer>();

-  optimizer->add_rule(std::make_unique<DependentGroupByReductionRule>());
+//  optimizer->add_rule(std::make_unique<DependentGroupByReductionRule>());

   optimizer->add_rule(std::make_unique<ExpressionReductionRule>());

