diff --git a/src/lib/optimizer/optimizer.cpp b/src/lib/optimizer/optimizer.cpp
index 916d2dbb6..488af44ee 100644
--- a/src/lib/optimizer/optimizer.cpp
+++ b/src/lib/optimizer/optimizer.cpp
@@ -128,7 +128,7 @@ std::shared_ptr<Optimizer> Optimizer::create_default_optimizer() {
   // Run the ColumnPruningRule before the PredicatePlacementRule, as it might turn joins into semi joins, which
   // can be treated as predicates and pushed further down. For the same reason, run it after the JoinOrderingRule,
   // which does not like semi joins (see above).
-  optimizer->add_rule(std::make_unique<ColumnPruningRule>());
+//  optimizer->add_rule(std::make_unique<ColumnPruningRule>());

   optimizer->add_rule(std::make_unique<SemiJoinReductionRule>());
